src/endless/Analytics.cpp

```diff
@@ -0,0 +1,207 @@
+}
+
+Analytics::~Analytics()
+{
+}
+
+Analytics *Analytics::instance()
+{
+	static Analytics instance;
+	return &instance;
+}
+
+void Analytics::sessionControl(BOOL start)
+{
+	if (m_disabled) return;
+	FUNCTION_ENTER;
+	CString body;
+	prefixId(body);
+	if (start) {
+		body = body + _T("t=screenview&cd=FirstPage&sc=start");
+		sendRequest(body);
+	}
+	else {
+		body = body + _T("t=screenview&cd=LastPage&sc=end");
+		sendRequest(body, TRUE);
+		DWORD ret = WaitForSingleObject(m_workerThread->m_hThread, 4000);
+		if (ret == WAIT_TIMEOUT) {
+			delete m_workerThread;
+			m_workerThread = NULL;
+		}
+	}
+}
+
+void Analytics::screenTracking(const CString &name)
+{
+	if (m_disabled) return;
+	FUNCTION_ENTER;
+	CString body, nameEnc;
+	prefixId(body);
+	urlEncode(name, nameEnc);
+	body.AppendFormat(_T("t=screenview&cd=%s"), nameEnc);
+	sendRequest(body);	
+}
+
+void Analytics::eventTracking(const CString &category, const CString &action, const CString &label, int value)
+{	
+	if (m_disabled) return;
+	FUNCTION_ENTER;
+	CString body, catEnc, actEnc;
+	prefixId(body);
+	urlEncode(category, catEnc);
+	urlEncode(action, actEnc);
+	body.AppendFormat(_T("t=event&ec=%s&ea=%s"), catEnc, actEnc);
+	if (!label.IsEmpty()) {
+		CString labelEnc;
+		urlEncode(label, labelEnc);
+		body.AppendFormat(_T("&el=%s"), labelEnc);
+	}
+	if (value >= 0) {
+		body.AppendFormat(_T("&ev=%d"), value);
+	}
+	sendRequest(body);
+}
+
+void Analytics::exceptionTracking(const CString &description, BOOL fatal)
+{
+	if (m_disabled) return;
+	FUNCTION_ENTER;
+	CString body, descEnc;
+	prefixId(body);
+	urlEncode(description, descEnc);
+	body.AppendFormat(_T("t=exception&exd=%s&exf=%d"), descEnc, fatal);
+	sendRequest(body);
+}
+
+void Analytics::setLanguage(const CString &language)
+{
+	m_language = language;
+}
+
+void Analytics::sendRequest(const CString &body, BOOL lastRequest)
+{
+	FUNCTION_ENTER;
+	CString *pBody = new CString(body);
+	//CWinThread *pThread = AfxBeginThread(threadSendRequest, pBody);
+	UINT msgType = WM_APP;
+	if (lastRequest) msgType = WM_APP+1;
+	m_workerThread->PostThreadMessage(msgType, (WPARAM)pBody, 0);
+}
+
+BOOL Analytics::disabled()
+{
+	CWinApp *pApp = AfxGetApp();
+	int disabled = pApp->GetProfileInt(_T(REG_SECTION), _T(REG_KEY_DISABLED), 0);
+	return (disabled == 1);
+}
+
+void Analytics::loadUuid(CString &uuid)
+{
+	CWinApp *pApp = AfxGetApp();
+	CString strUuid = pApp->GetProfileString(_T(REG_SECTION), _T(REG_KEY_CLIENT_ID));
+
+	if (strUuid.IsEmpty()) {
+		createUuid(strUuid);
+		pApp->WriteProfileString(_T(REG_SECTION), _T(REG_KEY_CLIENT_ID), strUuid);
+	}
+
+	uuid = strUuid;
+}
+
+void Analytics::createUuid(CString &uuid)
+{
+	const CString tokens = _T("0123456789abcdef");
+	srand((unsigned)time(NULL));
+	
+	// UUID v4 format xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx
+	uuid = _T("xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx");
+	int offset = 0;
+	for (int i = 0; i < 36; i++) {
+		if (i == 8 || i == 13 || i == 14 || i == 18 || i == 19 || i == 23)
```

Comment:
Why not replace this entire function with a call to [CoCreateGuid](https://msdn.microsoft.com/en-us/library/windows/desktop/ms688568(v=vs.85).aspx) which is used elsewhere? Is there a requirement that client IDs are v4 (random) UUIDs? I don't know what flavour CoCreateGuid creates, but… probably it's okay?